pkg/reconciler/finalizer/base_finalizer.go

```diff
@@ -56,3 +69,60 @@ func (bf *baseFinalizer) destroy(ctx context.Context) status.MultiError {
 	klog.Info("Destroyer succeeded")
 	return nil
 }
+
+func (bf *baseFinalizer) unmanageObjects(ctx context.Context, rsyncRef client.ObjectKey, hasSynced bool) status.MultiError {
+	var errs status.MultiError
+	rg := &v1alpha1.ResourceGroup{}
+	if err := bf.Client.Get(ctx, rsyncRef, rg); err != nil {
+		if apierrors.IsNotFound(err) {
+			if hasSynced {
+				// The ResourceGroup should exist if there has been a sync
+				klog.Error("ResourceGroup not found when attempting to unmanage objects for synced RSync")
+			} else {
+				// ResourceGroup doesn't exist, but that's expected since there hasn't been a sync
+				klog.Info("ResourceGroup not found when attempting to unmanage objects for unsynced RSync")
+			}
+			return nil
+		}
+		return status.APIServerError(err, "failed to get ResourceGroup", rg)
+	}
+	numNotFound := 0
+	for _, objRef := range rg.Spec.Resources {
+		// Remove metadata for each object
+		obj := &metav1.PartialObjectMetadata{}
+		obj.SetGroupVersionKind(schema.GroupVersionKind{
+			Group: objRef.Group,
+			Kind:  objRef.Kind,
+		})
+		key := client.ObjectKey{
+			Name:      objRef.Name,
+			Namespace: objRef.Namespace,
+		}
+		if err := bf.Client.Get(ctx, key, obj); err != nil {
+			if apierrors.IsNotFound(err) { // skip orphaning objects that don't exist
+				numNotFound++
+				continue
+			}
+			errs = status.Append(errs, status.APIServerError(err, "failed to get object", obj))
+			continue
+		}
+		existing := obj.DeepCopy()
+		// TODO: always remove apply set label with RemoveConfigSyncMetadata?
+		// All callers of RemoveConfigSyncMetadata also call RemoveApplySetPartOfLabel
+		update1 := metadata.RemoveConfigSyncMetadata(obj)
+		update2 := metadata.RemoveApplySetPartOfLabel(obj, bf.ApplySetID)
+		if update1 || update2 {
+			if err := bf.Client.Patch(ctx, obj, client.MergeFrom(existing),
+				client.FieldOwner(configsync.FieldManager)); err != nil {
+				errs = status.Append(errs, status.APIServerError(err, "failed to patch object to remove metadata", obj))
+			}
+		}
+	}
+	numErrs := 0
+	if errs != nil {
+		numErrs = len(errs.Errors())
+	}
+	klog.Infof("Finalizer unmanaged %d objects (%d not found) (%d errors)",
```

Comment:
Since this might be re-run multiple times, we might want to show progress (total, successfully unmanaged, not found, and errored), like we do in the Applier with SyncStats. There's even a `stats.DisabledObjStats` you might be able to re-use if you add a field for notfound. Then you can log it like: `klog.Infof("Finalizer made new progress: %s", unmanageStats)`. Might be good to populate an ObjectStatusMap too, so it's even more similar, and we can see exactly which objects fall into each category in the logs, for debugging. https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/pkg/parse/updater.go#L194-L198

Reply:
I ok with merging this as-is and having you rewrite the stats logging for both branches in a followup PR, if you really want to get this in for 1.21.0.

I'm also ok with merging it a day late, but we'll need Janet's approval.